src/tools/kotlin-type-generator.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * @license
+ * Copyright (c) 2020 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {KotlinGenerationUtils} from './kotlin-generation-utils.js';
+import {SchemaGraph, SchemaNode} from './schema2graph.js';
+import {HandleConnectionSpec} from '../runtime/particle-spec.js';
+import {Type} from '../runtime/type.js';
+import {HandleConnection} from '../runtime/recipe/handle-connection.js';
+import {assert} from '../platform/assert-web.js';
+import {generateSchema} from './kotlin-schema-generator.js';
+
+const ktUtils = new KotlinGenerationUtils();
+
+/**
+ * Generates a Kotlin type instance for the given handle connection.
+ */
+export async function generateConnectionType(connection: HandleConnection): Promise<string> {
+  return generateConnectionSpecType(connection.spec, new SchemaGraph(connection.particle.spec).nodes);
+}
+
+export async function generateConnectionSpecType(connection: HandleConnectionSpec, nodes: SchemaNode[]): Promise<string> {
+  let type = connection.type;
+  if (type.isEntity || type.isReference) {
+    // Moving to the new style types with explicit singleton.
+    type = type.singletonOf();
+  }
+
+  return generateType(type, type => {
+    if (!type.isEntity) return null;
+    if (connection.type.hasVariable) return null;
+    const node = nodes.find(n => n.schema.equals(type.getEntitySchema()));
+    return ktUtils.applyFun('EntityType', [`${node.fullName(connection)}.SCHEMA`]);
+  });
+}
+
+/**
+ * Generates a Kotlin representation of an Arcs type.
+ *
+ * @param overrideFunction optional lambda that allows overriding type generation at any place in the type hierarchy.
```

Comment:
Maybe we can expand the comment to state that returning null yields control over type creation.